src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -435,12 +431,19 @@ public function get($id, $invalidBehavior = ContainerInterface::EXCEPTION_ON_INV
 
         try {
             $definition = $this->getDefinition($id);
+            if ($definition instanceof ServiceAwareDefinition) {
+                return $definition->getService();
+            }
         } catch (ServiceNotFoundException $e) {
-            if (ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE !== $invalidBehavior) {
-                return;
+            if (ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE === $invalidBehavior) {
+                throw $e;
             }
 
-            throw $e;
+            return;
+        }
+
+        if (!$this->compiled) {
+            @trigger_error(sprintf('Calling %s() before compiling the container is deprecated for non-synthetic services since version 3.2 and will throw an exception in 4.0.', __METHOD__), E_USER_DEPRECATED);
```

Comment:
Updated message. Concerning BC: before this was triggered for each possible alias, now only once.